marshal/consumer.go

```diff
@@ -391,22 +423,21 @@ func (c *Consumer) isTopicClaimLimitReached(topic string) bool {
 // as the key, anybody who has that partition has claimed the entire topic. This requires all
 // consumers to use this mode.
 func (c *Consumer) claimTopics() {
-	// let's make a copy of what's in c.claimedTopics
-	latestClaims := make(map[string]bool)
-	if !c.Terminated() {
-		func() {
-			c.lock.RLock()
-			defer c.lock.RUnlock()
-			for topic, claimed := range c.claimedTopics {
-				latestClaims[topic] = claimed
-			}
-		}()
-	}
+	// Get a copy of c.partitions so we don't have to hold the lock throughout
+	// this entire method
+	topic_partitions := make(map[string]int)
+	func() {
+		c.lock.RLock()
+		defer c.lock.RUnlock()
 
-	for topic, partitions := range c.partitions {
+		for k, v := range c.partitions {
+			topic_partitions[k] = v
+		}
+	}()
+
+	// Now iterate each and try to claim
+	for topic, partitions := range topic_partitions {
 		if partitions <= 0 {
```

Comment:
How are you handling losing topics? This part existed to handle that case. We're comparing previous topics claimed against the current partition claims and remove things off if we no longer claim the topic.